test/common/quic/envoy_quic_h3_fuzz_helper.cc

```diff
@@ -0,0 +1,408 @@
+  } break;
+  case QuicFrame::kRetireConnectionId: {
+    // no retire frame in IETF mode
+    return;
+  } break;
+  case QuicFrame::kMaxStreams: {
+    const auto& f = frame.max_streams();
+    frames.push_back(quic::QuicFrame(
+        quic::QuicMaxStreamsFrame(f.control_frame_id(), f.stream_count(), f.unidirectional())));
+  } break;
+  case QuicFrame::kStreamsBlocked: {
+    const auto& f = frame.streams_blocked();
+    frames.push_back(quic::QuicFrame(
+        quic::QuicStreamsBlockedFrame(f.control_frame_id(), f.stream_count(), f.unidirectional())));
+  } break;
+  case QuicFrame::kPathResponse: {
+    const auto& f = frame.path_response();
+    frames.push_back(quic::QuicFrame(
+        quic::QuicPathResponseFrame(f.control_frame_id(), toPathFrameBuffer(f.data()))));
+  } break;
+  case QuicFrame::kPathChallenge: {
+    const auto& f = frame.path_challenge();
+    frames.push_back(quic::QuicFrame(
+        quic::QuicPathChallengeFrame(f.control_frame_id(), toPathFrameBuffer(f.data()))));
+  } break;
+  case QuicFrame::kStopSending: {
+    const auto& f = frame.stop_sending();
+    quic::QuicRstStreamErrorCode error_code =
+        static_cast<quic::QuicRstStreamErrorCode>(f.error_code());
+    frames.push_back(quic::QuicFrame(
+        quic::QuicStopSendingFrame(f.control_frame_id(), f.stream_id(), error_code)));
+  } break;
+  case QuicFrame::kMessageFrame: {
+    const auto& f = frame.message_frame();
+    size_t len = std::min(f.data().size(), sizeof(buffer));
+    memcpy(buffer, f.data().data(), len);
+    message_frame = std::make_unique<quic::QuicMessageFrame>(buffer, len);
+    frames.push_back(quic::QuicFrame(message_frame.get()));
+  } break;
+  case QuicFrame::kNewToken: {
+    const auto& f = frame.new_token();
+    size_t len = std::min(f.token().size(), sizeof(buffer));
+    memcpy(buffer, f.token().data(), len);
+    absl::string_view token(buffer, len);
+    new_token = std::make_unique<quic::QuicNewTokenFrame>(f.control_frame_id(), token);
+    frames.push_back(quic::QuicFrame(new_token.get()));
+  } break;
+  case QuicFrame::kAckFrequency: {
+    const auto& f = frame.ack_frequency();
+    auto delta = quic::QuicTime::Delta::FromMilliseconds(clampU64(f.milliseconds()));
+    ack_frequency = std::make_unique<quic::QuicAckFrequencyFrame>(
+        f.control_frame_id(), clampU64(f.sequence_number()), clampU64(f.packet_tolerance()), delta);
+    frames.push_back(quic::QuicFrame(ack_frequency.get()));
+  } break;
+  default:
+    return;
+  }
+  quic::QuicFramer framer({quic_version_}, connection_helper_->GetClock()->Now(),
+                          quic::Perspective::IS_CLIENT, quic::kQuicDefaultConnectionIdLength);
+
+  framer.SetEncrypter(quic::ENCRYPTION_INITIAL,
+                      std::unique_ptr<quic::QuicEncrypter>(new FuzzEncrypter()));
+  quic_packet_sizes_[idx_] =
+      framer.BuildDataPacket(header, frames, quic_packets_[idx_], sizeof(quic_packets_[idx_]),
```

Comment:
`BuildDataPacket()` with random frames could trigger DCHECK or ASSERT to cause the fuzzer to crash at this point. Have you run the fuzzer to seem if it's happening or not? If yes, you may need to do some input sanitation in above logic.

Reply:
Yes, I've run the fuzzer. Note that some frame types a deliberately no-op for this reason. Random as in random bytes ("junk") frames will completely bypass `BuildDataPacket()`.

Reply:
Can you comment about which frame type is skipped for hitting DCHECKs?

Reply:
Frames that I have included in the protobuf definition, which are actually not part of the IETF mode:

1. GO_AWAY
2. RETIRE_CONNECTION_ID
3. STOP_WAITING

I could probably remove them all together.

Reply:
You're right, RETIRE_CONNECTION_ID is an IETF frame. GOAWAY is not, at least according to quiche sources, c.f. quic_framer.cc:5296.

Reply:
GOAWAY frame is an HTTP/3 frame type rather than a QUIC frame type, see.  HttpEncoder::SerializeGoAwayFrame()

Reply:
Ah, I see. The H3 GoAway frame is being handled, too.